### ISSUE TITLE
Add an example that **doesn't use** the define_language macro?

### ISSUE THREAD

**Issue by sheredom** [egraphs-good/egg#219]
It'd be great if one of the tests (that you point to as examples of how to setup egg) didn't use the define_language macro - as I actually don't have a clue how to hook up egg to my toy compiler IR without having some sort of textual representation! I'm sure there _must_ be a way for me to hook up my toy IR to egg, I'm just not sure how!

**Comment by samuelgruetter** [comment]
I was missing such an example too! I eventually figured something out: I always use the predefined `SymbolLang`, and dynamically construct rewrites using `Rewrite::new`, and `.parse::<Pattern<SymbolLang>>().unwrap()` to construct a pattern from a string.
Not sure if that's what you're looking for, but here's some code I wrote:
https://github.com/samuelgruetter/egg/blob/344678d9465e8ac6ce76af132adf72caadf3c303/src/bin/coquetier.rs#L42

**Comment by mwillsey** [comment]
You can use any type you want! You just have to implement the [`Language`](https://docs.rs/egg/latest/egg/trait.Language.html) trait (and [`FromOp`](https://docs.rs/egg/latest/egg/trait.FromOp.html) if you want parsing and rust's `Display` if you want printing).

**Comment by sheredom** [comment]
How do I write `Rewrite`'s without using a textual format for them? Couldn't work that out either!

**Comment by mwillsey** [comment]
It's quite difficult to make `Rewrite`s without parsing. Note that you can implement `FromOp` yourself to get the ability to parse patterns. If you don't want to implement `FromOp`, then it you can manually create a [`PatternAst`](https://docs.rs/egg/latest/egg/type.PatternAst.html) can then you can call [`Pattern::new`](https://docs.rs/egg/latest/egg/struct.Pattern.html#method.new) on that. Then you can call [`Rewrite::new`](https://docs.rs/egg/latest/egg/struct.Rewrite.html#method.new) on a two patterns.

**Comment by sheredom** [comment]
Ok I'll try that ta!

**Comment by valarauca** [comment]
Seconding this request.

Trying to incorporate this crate into my existing DLS feels daunting to a near impossible task.

**Comment by mwillsey** [comment]
You might find the [implementation](https://docs.rs/egg/latest/src/egg/language.rs.html#828-840) of `Language` for the built-in `SymbolLang` somewhat helpful.

**Comment by valarauca** [comment]
I found that code sample, it wasn't.

**Comment by mwillsey** [comment]
Do you have a specific question? I am no longer actively working on the documentation, so I do not anticipate adding a full tutorial in the near future.

egg requires basically full control over your AST because of the way that the e-graph works, so there's not really a way to use the same type of tree-like `enum` that you'd use in a "conventional" compiler. Even if you don't use `define_language!`, you'll still have to completely convert your AST into the egg-friendly type that implements `Language`.

**Comment by robert-chiniquy** [comment]
Hi there @mwillsey!

I hope it's ok that I comment onto this issue, it seems very close to what I've just been working on figuring out. My goal is to use Egg with my own IR, which means not using the Sexp parser in Egg, (ideally, not using strings at all with Egg) which means not using the macros/methods which rely on strings/Sexp.

I thought that `define_language!` would probably be hard. Turns out it looks pretty easy though! Using rust-analyzer's "Expand macro recursively" feature gives me very comprehensible code. 

My only question about `define_language!` is pretty trivial: can I get away with not implementing `FromOp` for my language if I don't care about converting to/from strings? (I think the answer here is Yes)

I did run into something that I don't see how to support though, working through the rest of it– how to build my own rewrite rules without using strings. 

Starting with this example: https://github.com/egraphs-good/egg/blob/main/tests/prop.rs#L76

`rule! {def_imply, def_imply_flip,   "(-> ?a ?b)",       "(| (~ ?a) ?b)"          }`

"Expand macro recursively" gives (in part):
```rs
fn def_imply() -> Rewrite {
    {
        let searcher = "(-> ?a ?b)".parse::<$crate::Pattern<_>>().unwrap();
        let core_applier = "(| (~ ?a) ?b)".parse::<$crate::Pattern<_>>().unwrap();
        let applier = core_applier;
        $crate::Rewrite::new(("def_imply").to_string(), searcher, applier).unwrap()
    }
}
```

So now, to unwind the construction of types via parsing from strings, from that I get:
```rs
fn def_imply() -> Rewrite {
    let searcher: Pattern<Prop> = Pattern::new(RecExpr::<_>::from(vec![
        ENodeOrVar::Var(Var::from_str("?a").unwrap()),
        ENodeOrVar::Var(Var::from_str("?b").unwrap()),
        ENodeOrVar::ENode(Prop::Implies([0.into(), 1.into()])),
    ]));

    let applier: Pattern<Prop> = Pattern::new(RecExpr::<_>::from(vec![
        ENodeOrVar::Var(Var::from_str("?a").unwrap()),
        ENodeOrVar::ENode(Prop::Not(0.into())),
        ENodeOrVar::Var(Var::from_str("?b").unwrap()),
        ENodeOrVar::ENode(Prop::Or([1.into(), 2.into()])),
    ]));

    egg::Rewrite::new(("def_imply").to_string(), searcher, applier).unwrap()
}
```

Which is not bad at all, barely more verbose than using the macro IMO. However, in the above, strings are still used to define a `Var` for use in `ENodeOrVar::Var`.

`ENodeOrVar` is generic over the language but requires the concrete type `Var` for the variant `ENodeOrVar`. 
```rs
pub enum ENodeOrVar<L> {
    /// An enode from the underlying [`Language`]
    ENode(L),
    /// A pattern variable
    Var(Var),
}
```

The type `Var` (`pub struct Var(Symbol);`) takes a `Symbol` (`pub use symbol_table::GlobalSymbol as Symbol;`), which as `GlobalSymbol` wraps a `Symbol` (`pub struct Symbol(NonZeroU32);`) which wraps a `u32`. 

The fields of these public types are not themselves public, which means I can't just hack them together, and aside from `from_str` I don't see any other way to construct one (which makes sense given that they are explicitly intended to index strings in an interned table).

It seems to me from reading the code that the only purpose of `Var`s is to act as unique identifiers, in which case I think a string-free API would make a lot of sense to me.

Solutions? 

Here are the only ideas I've come up with:
1. Maybe there is some constructor / pathway to build these types which I have overlooked?
2. `ENodeOrVar` could be generic over `V` and `ENodeOrVar::Var` could become `ENodeOrVar::Var(V)`, parallel to how `L` works there. This might not be as dramatic a change as it seems, the default type for the generic `V` could still be `Var`, so I would think all existing code would be fine, this would just let me use something else there, say `u32`.
3. `Var` could be simplified to simply carry a `u32` as a public field (or with some other constructor), allowing me to construct one manually using my own indexes (I can be responsible for their correctness)
4. ... I could make some system to serialize my IR as strings so I can intern them as symbols, this lets me just use egg as-is (but of course then, `GlobalSymbol`, which is intended to improve performance, dramatically degrades performance for me)
5. ... To keep it very simple, a method like `Var::from_fake_global_symbol(usize) -> Var` method could be added, which I could call passing in my own indexes (as in solution 3 above). (This is obviously not great but it is also how I would probably have to hack it if I were just going to fork upstream to get by for the time being)

So, any help / guidance / tips are welcome and appreciated. If any of the above solutions would be welcome as a PR, I would be happy to put the work it to contribute it, let me know.

Also, I love the library, I think the code is great, docs are great too, I love the associated papers, I think e-graphs will really help me do what I am trying to do, so thank you very much for all the hard work ❤️
